### PR TITLE
DEV: Added non_composable to enum, removed composable argument from d…

### PR DESCRIPTION
**Pull request by Nick-Foto**
…efine_app, excluded _add method for non_composable apps

### CODE REVIEW

**Comment by Nick-Foto** [comment]
@GavinHuttley 
Please review and merge these changes.
Please remind making a tag before merging this branch.

**Comment by Nick-Foto** [comment]
@GavinHuttley 
I changed algorithm for checking existing functions in class. Items in `method_list` and `__mapping` were different, so some of functions were not checked. Because `__str__` doesn't exist in `__mapping`, the previous code didn't check it.  
I added complete tests for this part of code, also tests for `_add` and `min_params` in `_get_raw_hints`.

File: src/cogent3/app/composable.py [hunk 1]
```diff
@@ -848,15 +850,17 @@ def _disconnect(self):
 
 
 def _add(self, other):
-    if other.input is not None:
+    if hasattr(other, "input") and other.input is not None:
```
Comment (GavinHuttley) [inline src/cogent3/app/composable.py:853]: Remove this `hasattr` check. Instead add a new conditional before this
```python
if getattr(other, `app_type`, None) not in {WRITER, LOADER, GENERIC}:
    raise TypeError(f"{other!r} is not composable"
```

File: src/cogent3/app/composable.py [hunk 2]
```diff
@@ -848,15 +850,17 @@ def _disconnect(self):
...
         raise ValueError(
             f"{other.__class__.__name__} already part of composed function, use disconnect() to free them up"
         )
     if other is self:
         raise ValueError("cannot add an app to itself")
 
     # Check order
-    if self.app_type is WRITER:
+    if self.app_type is NON_COMPOSABLE or other.app_type is NON_COMPOSABLE:
```
Comment (GavinHuttley) [inline src/cogent3/app/composable.py:861]: delete this conditional,  addressed above.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
I suggest defining a new variable within the `define_app` function to simplify the expressions
```python
composable = app_type is not NON_COMPOSABLE
```
then you can just revert all those statements to using the composable variable making them easier to read

File: src/cogent3/app/composable.py (hunk 1 above)
Comment (Nick-Foto) [reply src/cogent3/app/composable.py:853]: Why for the first condition you used getattr, but for the next, directly .app_type?

**Pull request merged**